### PR TITLE
Topic Recognition - StyleGAN1 Implementation - Christopher Atkinson - 44776859

### PR DESCRIPTION
**Pull request by Catkinsv3**
Hi Shakes (and COMP3710 team),

Please review my pull request containing an implementation of StyleGAN1 as applied to the OAI AKOA Knee dataset and OASIS Brain dataset. The model was trained independently on each of these datasets, along with MNIST Digits and CelebA Faces datasets, the results of which are discussed in the README file and displayed in video form.

- **model.py**: contains functionality to ingest and pre-process the dataset, build the models and execute training.
- **driver.ipynb**: contains test scripts that train the model on each of the 4 datasets mentioned, after initialising the model with appropriate hyperparameters. Please ensure you update the paths to your local copies of the datasets you wish to test on.
- **README.md**: A report on the work done so far, and a demonstration and discussion of the results.
- **resources folder**: Contains images used in the README file.

Please let me know if any aspect of this project could be improved or expanded upon, or if any issues need to be resolved.

Kindest regards,
Christopher Atkinson
44776859

### CODE REVIEW

**Comment by WeiDaiwow** [comment]
1. Great code with great comments
2. Nice readme file with great explanation
3. Very nice results on different datasets

Suggestion:
The same output for the OASIS dataset after 20 epochs shows that the discriminator is too powerful compared to the generator for this dataset. This is also shown in your balance spectrum that unbalanced training is definitely happening. One approach you can take is to reduce the capacity of the discriminator by using any kind of dropout layers to reduce the capacity of it. The less information discriminator can see during the training, the more uncertainty there is.

**Review by WeiDaiwow — COMMENTED** [review]
Great code with great comments
Nice readme file with great explanation
Very nice results on different datasets
Suggestion:
The same output for the OASIS dataset after 20 epochs shows that the discriminator is too powerful compared to the generator for this dataset. This is also shown in your balance spectrum that unbalanced training is definitely happening. One approach you can take is to reduce the capacity of the discriminator by using any kind of dropout layers to reduce the capacity of it. The less information discriminator can see during the training, the more uncertainty there is.

**Comment by Catkinsv3** [comment]
Thank you for the feedback! Increasing the uncertainty is an interesting idea. I did experiment with adding batch normalisation and dropout to the discriminator, both individually and together, but there was no observable improvement unfortunately (perhaps I did not use the right dropout amount). I had some success with limiting the neurons/filters in the discriminator, but limiting it more than it is now allows the generator to become too dominant. I'm sure that good results could be achieved with further tuning, with some combination of learning rate, dropout and network size, if there were time. Thanks again.

**Comment by shakes76** [comment]
## TF/Torch Usage
Adequate use and implementation
## Good Practice (Design/Commenting)
Good spacing and comments
Header blocks
## Algorithm 
Driver Script present
Shows Usage & Demo & Visualisation & Data usage 
Module present 
No Data leakage
Difficulty: Hard
## Commit Log
Meaningful commit messages
Progressive commits used
## Documentation
Good Description and Comments 
Markdown used PDF submitted
## Pull Request
Successful Pull Request (Working Algorithm Delivered on Time in Correct Branch)
No feedback required
Request Description good
